File: C++/Exceptions/WithoutExceptionsPerformance.cpp
#include <iostream>
#include "ExceptionsBenchmark.cpp"

namespace oobench {

    class WithoutExceptionsPerformance : public ExceptionsBenchmark {
    public:
        WithoutExceptionsPerformance(int argc, const char** argv) 
                : ExceptionsBenchmark(argc, argv) { 
        }

        virtual int getMinorNumber() const {
            return 1;
        }

        static int simpleException() {
            return 1;
        }

        static int deepException2() {
            return 1;
        }

        static int deepException() {
            int result = deepException2();
            if (1 == result) {
                return 0;
            } else if (2 == result) {
                return 0;
            } 
            return result;
        }

        virtual void except(unsigned long items) {
            for (unsigned long i = 0; i < items; ++i) {
                if (simpleException() == 1) {
                }
            }
        }

        virtual void exceptDeep(unsigned long items) {
            for (unsigned long i = 0; i < items; ++i) {
                if (deepException() == 1) {
                }
            }
        }

    };
}

int main(int argc, const char** argv) {
    SHOW_LOCATION("Exceptions");
    try {
        oobench::WithoutExceptionsPerformance performance(argc, argv);
        performance.test(500000L, "without exceptions");
    } catch (...) {
        std::cerr << "Unknown exception caught" << std::endl;
    }
    return 0;
    std::exit(0);
}

File: Objective-C/Foundation/Exceptions/WithoutExceptionsPerformance.m
#include <Foundation/Foundation.h>
#include "ExceptionsBenchmark.m"

@interface WithoutExceptionsPerformance: ExceptionsBenchmark
@end

@implementation WithoutExceptionsPerformance

- (int)minorNumber
{
    return 1;
}

+ (int)simpleException
{
    return 1;
}

+ (void)exceptWithCount:(unsigned long)count
{
    unsigned long i = 0;

    for (i = 0; i < count; i++) {
        if ([self simpleException] == 1) {
        }
    }
}

+ (int)deepException2
{
    return 3;
}

+ (int)deepException
{
    long result = [self deepException2];

    if (1 == result) {
        return 0;
    } else if (2 == result) {
        return 0;
    }
    return result;
}

+ (void)exceptDeepWithCount:(unsigned long)count
{
    unsigned long i = 0;

    for (i = 0; i < count; i++) {
        if ([self deepException] > 0) {
        }
    }
}

@end

int main(int argc, char **argv, char **env) {
    SHOW_LOCATION("Foundation/Exceptions");
#if LIB_FOUNDATION_LIBRARY
    [NSProcessInfo initializeWithArguments:argv count:argc
            environment:env];
#endif
    {
        OOB_CREATE_AUTORELEASE_POOL(pool);
        WithoutExceptionsPerformance *performance =
                [WithoutExceptionsPerformance 
                newBenchmarkWithArguments:argc : argv];

        [performance testWithCount:500000 description:@"without exceptions"];
        OOB_RELEASE(pool);
    }
    exit(0);
    return 0;
}
